utility/src/main/java/org/oppia/util/parser/LiTagHandler.kt

```diff
@@ -1,37 +1,171 @@
+     */
+    fun closeItem(text: Editable, indentation: Int)
+  }
+
+  /**
+   * Subclass of [ListTag] for unordered lists.
+   */
+  private inner class Ul : ListTag {
+
+    override fun openItem(text: Editable) {
+      appendNewLine(text)
+      start(text, BulletListItem())
+    }
+
+    override fun closeItem(text: Editable, indentation: Int) {
+      appendNewLine(text)
+
+      getLast<BulletListItem>(text)?.let { mark ->
+        setSpanFromMark(text, mark, CustomBulletSpan(context, indentation, "•"))
       }
     }
   }
+
+  /**
+   * Subclass of [ListTag] for ordered lists.
+   */
+  private inner class Ol : ListTag {
+
+    private var index = 1
+
+    override fun openItem(text: Editable) {
+      appendNewLine(text)
+      start(text, NumberListItem(index))
+      index++
+    }
+
+    override fun closeItem(text: Editable, indentation: Int) {
+      appendNewLine(text)
+      getLast<NumberListItem>(text)?.let { mark ->
+        setSpanFromMark(text, mark, CustomBulletSpan(context, indentation, "${mark.number}."))
+      }
+    }
+  }
+
+  /**
+   * These static methods are based on the Android Html class source code.
+   * https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/text/Html.java
+   */
+  companion object {
+
+    /**
+     * Appends a new line to [text] if it doesn't already end in a new line
+     */
+    private fun appendNewLine(text: Editable) {
+      if (text.isNotEmpty() && text.last() != '\n') {
+        text.append("\n")
+      }
+    }
+
+    /**
+     * Returns the most recently added span of type [T] in [text].
+     *
+     * Invisible marking spans are inserted to record the location of opening HTML tags in the text.
+     * We do this rather than using a stack in case text is inserted and the relative location shifts around.
+     *
+     * The last span corresponds to the top of the "stack".
+     */
+    private inline fun <reified T : Mark> getLast(text: Spanned) =
+      text.getSpans(0, text.length, T::class.java).lastOrNull()
+
+    /**
+     * Pops out the invisible [mark] span and uses it to get the opening tag location.
+     * Then, sets a span from the opening tag position to closing tag position.
+     */
+    private fun setSpanFromMark(text: Spannable, mark: Mark, styleSpan: Any) {
+      // Find the location where the mark is inserted in the string.
+      val markerLocation = text.getSpanStart(mark)
+      // Remove the mark now that the location is saved
+      text.removeSpan(mark)
+
+      val end = text.length
+      if (markerLocation != end) {
+        text.setSpan(styleSpan, markerLocation, end, SPAN_EXCLUSIVE_EXCLUSIVE)
+      }
+    }
+
+    /**
+     * Inserts an invisible [mark] span that doesn't do any styling.
+     * Instead, [setSpanFromMark] will later find the location of this span so it knows where the opening tag was.
+     */
+    private fun start(text: Spannable, mark: Mark) {
```

Comment:
Out of curiosity, why not just store the position  in the list tag object instead of adding a temporary mark? Is it because we would need to factor in the positions of later custom tags that might be added?

Reply:
Yes actually we do not know where to put the span using start and end indexes. openItem is too early and we don’t know where the closing tag is. But if we try to insert the span in closeItem, we don’t know where the span started anymore. This marker will be used to find the location of the opening tag inside the closeItem method.

Reply:
Thanks. can you add a bit more context to the KDoc above since that might be a bit subtle (hence my question--if I had gone and changed this without realizing the important of that behavior, things would have broken). Please also make sure your tests cover this case. (E.g. would they fail if the code was changed as I asked in my original question above?)

Reply:
I have added more details to Kdoc.